### PR TITLE
Initial merge with xeus-clang-repl - Capture outputs

### PR DESCRIPTION
**Pull request by alexander-penev**

### CODE REVIEW

File: src/xinterpreter.cpp [hunk 1]
```diff
@@ -104,6 +105,7 @@ __get_cxx_version ()
         createInterpreter(Args(argv ? argv + 1 : argv, argv + argc));
         m_version = get_stdopt();
         redirect_output();
+        // Bootstrap the execution engine
```
Comment (vgvassilev) [inline src/xinterpreter.cpp:108]: We should probably drop this line.

File: src/xinterpreter.cpp [hunk 2]
```diff
@@ -175,8 +188,13 @@ __get_cxx_version ()
             errorlevel = 1;
             ename = "Error :";
             evalue = "Compilation error!";
+            //evalue = error_stream.str(); //?
```
Comment (vgvassilev) [inline src/xinterpreter.cpp:191]: And this one.

File: src/xinterpreter.cpp [hunk 3]
```diff
@@ -175,8 +188,13 @@ __get_cxx_version ()
...
+            std::cerr << err;
         }
 
+        //error_stream.str().clear(); //?
+        //DiagnosticsOS.str().clear(); //?
```
Comment (vgvassilev) [inline src/xinterpreter.cpp:196]: And these.

File: src/xinterpreter.cpp [hunk 4]
```diff
@@ -1,5 +1,6 @@
 /************************************************************************************
  * Copyright (c) 2023, xeus-cpp contributors                                        *
+ * Copyright (c) 2023, Johan Mabille, Loic Gouarin, Sylvain Corlay, Wolf Vollprecht *
```
Comment (vgvassilev) [inline src/xinterpreter.cpp:3]: We should remove that line, too.

File: src/xinterpreter.cpp [hunk 5]
```diff
@@ -59,6 +59,20 @@ using namespace std::placeholders;
 
 namespace xcpp
 {
+    struct StreamRedirectRAII {
+      std::string *err = NULL;
```
Comment (vgvassilev) [inline src/xinterpreter.cpp:63]:
```suggestion
      std::string &err;
```

File: src/xinterpreter.cpp [hunk 6]
```diff
@@ -59,6 +59,20 @@ using namespace std::placeholders;
...
+      StreamRedirectRAII(std::string *e) {
+        err = e;
```
Comment (vgvassilev) [inline src/xinterpreter.cpp:65]:
```suggestion
      StreamRedirectRAII(std::string &e) : err(e) {
```

File: src/xinterpreter.cpp [hunk 7]
```diff
@@ -59,6 +59,20 @@ using namespace std::placeholders;
...
+        Cpp::BeginStdStreamCapture(Cpp::kStdErr);
+        Cpp::BeginStdStreamCapture(Cpp::kStdOut);
+      }
+      ~StreamRedirectRAII() {
+        std::string out = Cpp::EndStdStreamCapture();
+        *err = Cpp::EndStdStreamCapture();
+        std::cout << out;
```
Comment (vgvassilev) [inline src/xinterpreter.cpp:72]: We should probably print out where `err` is printed...

**Comment by anutosh491** [comment]
Hey @alexander-penev could you maybe squash the commits into a couple meaningful ones. Would make it easier for reviewing !

**Comment by vgvassilev** [comment]
> Hey @alexander-penev could you maybe squash the commits into a couple meaningful ones. Would make it easier for reviewing !

Often squashing loses the previous comments. Maybe we can squash on merge. If you click on the “files changed” tab you will see everything squashed.

**Review by JohanMabille — COMMENTED** [review]
LGTM, but I still don't understand why this is needed. Are `kStdErr` and `kStdOut` different beasts from `std::cerr` and `std::cout` ? Also, why capturing / restoring them on each execute request instead of doing it upon creation / destruction of the interpreter?

**Review by JohanMabille — APPROVED** [review]

**Pull request merged**